packages/@aws-cdk/aws-stepfunctions/lib/states/distributed-map.ts

```diff
@@ -0,0 +1,801 @@
+ */
+export class S3ManifestItemReader extends S3ItemReader {
+
+  protected readonly inputType: string = 'MANIFEST';
+
+}
+
+/**
+ * Interface for Result Writer configuration properties
+ */
+export interface ResultWriterProps {
+  /**
+   * S3 Bucket in which to save Map Run results
+   */
+  readonly bucket: IBucket
+
+  /**
+   * S3 prefix in which to save Map Run results
+   *
+   * @default - No prefix
+   */
+  readonly prefix?: string
+}
+
+/**
+ * Configuration for writing Distributed Map state results to S3
+ */
+export class ResultWriter {
+
+  /**
+   * S3 Bucket in which to save Map Run results
+   */
+  readonly bucket: IBucket
+
+  /**
+   * S3 prefix in which to save Map Run results
+   *
+   * @default - No prefix
+   */
+  readonly prefix?: string
+
+  constructor(props: ResultWriterProps) {
+    this.bucket = props.bucket;
+    this.prefix = props.prefix;
+  }
+
+  /**
+   * Render ResultWriter in ASL JSON format
+   */
+  public render(): any {
+    return FieldUtils.renderObject({
+      Resource: 'arn:aws:states:::s3:putObject',
+      Parameters: {
+        Bucket: this.bucket.bucketName,
+        ...(this.prefix && { Prefix: this.prefix }),
+      },
+    });
+  }
+
+  /**
+   * Compile policy statements to provide relevent permissions to the state machine
+   */
+  public providePolicyStatements(): iam.PolicyStatement[] {
+
+    const resource = `arn:aws:s3:::${this.bucket.bucketName}/*`;
+
+    return [
+      new iam.PolicyStatement({
+        actions: [
+          's3:PutObject',
+          's3:GetObject',
+          's3:ListMultipartUploadParts',
+          's3:AbortMultipartUpload',
+        ],
+        resources: [resource],
+      }),
+    ];
+  }
+}
+
+/**
+ * Percentage of failed items to tolerate in a Map Run
+ */
+export class ToleratedFailurePercentage {
+
+  /**
+   * Fill value from JSON path
+   */
+  public static fromPath(path: string): ToleratedFailurePercentage {
+    return new ToleratedFailurePercentage(path, true);
+  }
+
+  /**
+   * Fill value from number
+   */
+  public static fromNumber(value: number): ToleratedFailurePercentage {
+    return new ToleratedFailurePercentage(value);
+  }
+
+  /**
+   * Number value or string JSON path
+   */
+  private readonly value: any;
+
+  /**
+   * Is the value a string JSON path
+   */
+  private readonly fromPath?: boolean
+
+  private constructor(value: any, fromPath?: boolean) {
+    this.value = value;
+    this.fromPath = fromPath;
+  }
+
+  /**
+   * Renders the value as ASL JSON
+   */
+  public render(): any {
+    return {
+      [(this.fromPath ? 'ToleratedFailurePercentagePath' : 'ToleratedFailurePercentage')]: this.value,
+    };
+  }
+}
+
+/**
+ * Number of failed items to tolerate in a Map Run
+ */
+export class ToleratedFailureCount {
+
+  /**
+   * Fill value from JSON path
+   */
+  public static fromPath(path: string): ToleratedFailureCount {
```

Comment:
Please add a test using this method

Reply:
How fine-grained do unit tests need to be? I created a unit test that does a combination of `fromNumber` and `fromPath` for the properties here that all follow the same pattern. That way I have a single test that covers the various combinations of the pattern without necessarily covering the specific methods of each individual property using that pattern. So for example, there is a single test that calls `ToleratedFailurePercentage.fromPath` and `ToleratedFailureCount.fromNumber` to make sure both patterns are covered even though they are not covered for each object that uses that pattern.

Reply:
We prefer one test to cover just one test case. When one test covers too much functionality, it makes troubleshooting future errors much more difficult.

Reply:
I get that approach, but it also seems somewhat excessive to have an individual test for each possible parameter value.  `fromPath` is being tested and `fromNumber` is being tested.  These are very simple code blocks that would be very easy to troubleshoot errors in a single test.  I totally agree with separating things into multiple tests when it or more complex logic that will be difficult to troubleshoot and are unrelated.